Implement predicate pruning for `like` expressions (prefix matching)

adriangb opened the pull request:
The idea is that we can push certain `like` expressions down into statistics pruning.
For example, a filter like `url LIKE 'https://www.google.com%'` can be (basically, some caveats and other tricks) used to filter `url_min <= 'https://www.google.com' and 'https://www.google.com' <= url_max` such that a row group that only has `https://www.example.com` would be excluded.

Closes https://github.com/apache/datafusion/issues/507
Closes https://github.com/apache/datafusion/pull/13253

adriangb commented:
cc @alamb

alamb commented:
This is very clever -- I will review it tomorrow

alamb commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1487,6 +1489,17 @@ fn build_predicate_expression(
+            let op = match (like_expr.negated(), like_expr.case_insensitive()) {
+                (false, false) => Operator::LikeMatch,
+                (true, false) => Operator::NotLikeMatch,
+                (false, true) => Operator::ILikeMatch,
+                (true, true) => Operator::NotILikeMatch,
```
I think this is dead code as `if like_expr.case_insensitive() {` catches the case insensitive case

I think this code would be clearer if it just matched on `like_expr.negated()` (or alternately returned `unhandled_hook.handle(expr);` directly for these last 2 cases

adriangb replied:
Sure will change it to do so. I think I was getting a bit ahead of myself to implement `ILIKE` support, which as per the comment should be possible, maybe you can show me how to construct the physical expression to call `lower()` and `upper()` on another expression.

alamb commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1610,6 +1625,93 @@ fn build_statistics_expr(
 
+fn extract_string_literal(expr: &Arc<dyn PhysicalExpr>) -> Result<&String> {
+    if let Some(lit) = expr.as_any().downcast_ref::<phys_expr::Literal>() {
+        if let ScalarValue::Utf8(Some(s)) = lit.value() {
```
I think you should probably also handle the cases `ScalarValue::LargeUtf8`, `ScalarValue::Utff8View` as well.

adriangb replied:
And `Dictionary`!

alamb commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1610,6 +1625,93 @@ fn build_statistics_expr(
+    if prefix.is_empty() {
+        return plan_err!("Empty prefix in LIKE expression");
+    }
+    Ok(Arc::new(phys_expr::Literal::new(ScalarValue::Utf8(Some(
```
Do we also have to test if there are other occurences of `%` in the string 🤔  (like `foo%bar%`)?

adriangb replied:
The logic is pretty simple (truncate at the first one) but I agree another test would be nice.

alamb commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1610,6 +1625,93 @@ fn build_statistics_expr(
+    // column LIKE '%foo%' => min <= '' && '' <= max => true
+    // column LIKE 'foo' => min <= 'foo' && 'foo' <= max
+
+    // I *think* that ILIKE could be handled by making the min lowercase and max uppercase
```
👍 I agree. Figuring out how to make those call would be the trick

alamb commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1610,6 +1625,93 @@ fn build_statistics_expr(
+fn build_like_match(
+    expr_builder: &mut PruningExpressionBuilder,
+) -> Result<Arc<dyn PhysicalExpr>> {
+    // column LIKE literal => (min, max) LIKE literal split at % => min <= split literal && split literal <= max
```
I don't understand the `LIKE literal split at %` part 

`column LIKE literal` is the same as `column = literal` if there are no wild cards, so you should be able to use the same rules as equality I think

adriangb replied:
Right that's the point, by splitting it at the first `%` we are able to apply the same rules as equality:
```
column LIKE literal -> (min, max) LIKE (literal split at %) -> min <= split literal && split literal <= max
vs
column = literal -> (min, max) = literal -> min <= literal && literal <= max
```

alamb commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -3443,6 +3545,115 @@ mod tests {
+        let expected_ret = &[true, true, false, false, true, true];
+
+        prune_with_expr(
+            // s1 LIKE 'A%'
```
the comments say s1 LIKE `A%` but the code builds a different expressions

adriangb replied:
Okay I've tried to get all of these comments right, I think they can be removed tbh the expression is pretty self explanatory, but left them for now, let me know if you'd prefer that I remove them or if you want to keep them if there are any obviously wrong

alamb commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -3443,6 +3545,115 @@ mod tests {
+        let expected_ret = &[true, true, false, false, true, true];
+
+        prune_with_expr(
+            // s1 LIKE 'A%'
```
Can you also please add tests for other combinations:
* `s1 LIKE 'A'`
* `s1 LIKE '%A%'`
I think it is important the matching is doing the right thing

I also think it is important to to cover cases for NOT LIKE as well

alamb commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -3443,6 +3545,115 @@ mod tests {
+        let expected_ret = &[true, true, true, true, true, true];
+
+        prune_with_expr(
+            // s1 LIKE 'A%'
```
```suggestion
            // s1 LIKE '%A'
```

alamb replied:
this one is still wrong

adriangb replied:
Yeah I haven't pushed pending discussing the correctness of the general approach.

alamb reviewed (commented):
Thank you @adriangb  -- this is very cool. 

I think there are a few more tests needed but otherwise the implementation looks very solid. Thank you

adriangb commented:
@alamb can these stats be truncated? I know stats in pages truncate large strings, e.g. if the min value is `"B"` could it be that the actual min value is `"BA"`? If so I think this approach may not work at all. Imagine we have a row group with data `["BA", "ZD"]` which generates min/max stats `["B", "Z"]`. Now we want to know if `col LIKE '%A%'` is possible. Clearly the answer should be *yes* but if we convert it to the predicate form we get `'B' <= '' AND '' <= 'Z'` which gives `false` 😞

alamb commented:
@adriangb  in theory I think parquet statistics can be truncated. 

> Now we want to know if col LIKE '%A%'

I don't think we can use statistics for substring match -- we can only use statistics for equality and prefix matching

so like `col LIKE 'A%'`

The predicate would be transformed into `'B' <= 'A' AND 'A' <= 'Z'` which I do think is correct

adriangb commented:
Consider the values `["ABC", "XYZ"]` with stats `["AB", "XY"]` and the filter `col like 'A%'`. This becomes `'AB' <= 'A' AND 'A' <= 'XY'` which is false, but we need true. To fix this we'd need to truncate the stats to the length of the filter to get `'A' <= 'A' AND 'A' <= 'XY'` which then gives the right result. `%A%` is just an obvious case because you get `''` as the prefix which gives obvious issues.

adriangb commented:
Okay @alamb I pushed a pretty big rework. Lots of new test cases, lots of comments explaining what's going on. I removed the `not like` part; I'm thinking this is complex enough as is and most of the benefit (maybe even in ClickBench?) will come from `like`. We can tackle `not like`, `ilike` and `not ilike` in future PRs. Especially since it's going to be important to evaluate each test case carefully.

I will note that I am a bit concerned about the interaction of truncated stats and how we apply these filters. Take the (absurd) case of stats that were truncated so that all you have is `"",""`. You basically know _nothing_ about the data, there could be anything in there. Yet `col = 'A'` transforms into `'' <= 'A' and 'A' <= ''` which is false. Substitute in non-truncated stats and `'A' <= 'A' and 'A' <= 'Z'` is true. I added a test case for this behavior on the existing code. It doesn't differentiate between `"ABC"` truncated to `""` and `""` actually being the min string but it shows the behavior which would be the same in both cases.

adriangb commented:
Argh the only ClickBench query this maybe could improve is [23](https://github.com/ClickHouse/ClickBench/blob/d592b3d2dd59a5b2c887b4dc4f94b3ad52e6d5d6/datafusion/queries.sql#L23):
```sql
WHERE "Title" LIKE '%Google%' AND "URL" NOT LIKE '%.google.%'
```
Since the `like` starts with a wildcard this won't help. Maybe we can get `not like` to do something smart there, tbd...

Dandandan commented:
I am wondering if simple `like` patterns are not already converted to `startswith` etc, such that pruning already is applied.

adriangb commented:
> I am wondering if simple `like` patterns are not already converted to `startswith` etc, such that pruning already is applied.

Good question. Based on performance of some queries I saw I'd say no, but it it's worth double checking. Any suggestions as to a definitive easy way to check? I guess I can run `datafusion-cli` against a very large parquet file in object storage (high latency) and a query that should filter (`col like 'A%'`) and one that can't?

I don't see where `startswith` or any other transformations (`lower`, `upper`, etc.) are handled in the pruning transformation.

Dandandan commented:
I would take a look at the (physical plan) of queries involving like first to see if it still uses like or is transformed to another function.

adriangb commented:
I made a big parquet file as follows:

```python
import random
import string
import polars as pl

df = pl.DataFrame({'col': ["A" + "".join(random.choices(string.ascii_letters, k=1_000)) for _ in range(1_000_000)]})
df.write_parquet('data.parquet', compression='uncompressed')
```

This came out to ~1GB. I then uploaded it to a GCS bucket.

I ran queries `col = 'Z'` and `col like 'Z'` against it and got 2s and 23s respectively. IMO that means it's not getting pushed down.

The explain plans show the following:

```
ParquetExec: file_groups={10 groups: [[data.parquet:0..100890471], [data.parquet:100890471..201780942], [data.parquet:201780942..302671413], [data.parquet:302671413..403561884], [data.parquet:403561884..504452355], ...]}, projection=[col], predicate=col@0 = Z, pruning_predicate=CASE WHEN col_null_count@2 = col_row_count@3 THEN false ELSE col_min@0 <= Z AND Z <= col_max@1 END, required_guarantees=[col in (Z)], metrics=[output_rows=0, elapsed_compute=10ns, predicate_evaluation_errors=0, bytes_scanned=19368790, row_groups_pruned_bloom_filter=0, row_groups_pruned_statistics=3, pushdown_rows_filtered=0, page_index_rows_filtered=0, row_groups_matched_statistics=0, row_groups_matched_bloom_filter=0, file_scan_errors=0, file_open_errors=0, num_predicate_creation_errors=0, time_elapsed_scanning_until_data=18.748µs, time_elapsed_opening=7.717746249s, time_elapsed_processing=64.457827ms, page_index_eval_time=10.134µs, pushdown_eval_time=20ns, time_elapsed_scanning_total=19.21µs]
```

```
ParquetExec: file_groups={10 groups: [[data.parquet:0..100890471], [data.parquet:100890471..201780942], [data.parquet:201780942..302671413], [data.parquet:302671413..403561884], [data.parquet:403561884..504452355], ...]}, projection=[col], predicate=col@0 LIKE Z, metrics=[output_rows=1000000, elapsed_compute=10ns, predicate_evaluation_errors=0, bytes_scanned=1006955145, row_groups_pruned_bloom_filter=0, row_groups_pruned_statistics=0, pushdown_rows_filtered=0, page_index_rows_filtered=0, row_groups_matched_statistics=0, row_groups_matched_bloom_filter=0, file_scan_errors=0, file_open_errors=0, num_predicate_creation_errors=0, time_elapsed_scanning_until_data=49.346124581s, time_elapsed_opening=2.18377s, time_elapsed_processing=1.545583231s, page_index_eval_time=20ns, pushdown_eval_time=20ns, time_elapsed_scanning_total=49.654700084s]
```

alamb commented:
> So it doesn't seem like it's being transformed into another expression. It probably would be smart to do so as a general optimization outside of pruning.

> I also think pruning should handle whatever that produces (startswith in the case of like 'A%' or = in the case of like 'A') as well as additional simple cases like upper(), lower(), etc.

That certainly makes sense to me

Perhaps we could implement some sort of simplification in https://github.com/apache/datafusion/blob/main/datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs for `LIKE` into `starts_with` https://docs.rs/datafusion/latest/datafusion/functions/expr_fn/fn.starts_with.html (though since it is a function figuring out how to make the rewrite work might be tricky)

Then we can implement the rules in pruning predicate for starts_with 🤔

adriangb commented:
I'm happy to do it that way if you'd like. Or we can leave that for a future optimization.

Do you have any thoughts on my concerns for possibly truncated stats, in particular how `=` may even be wrong as of today if the stats are truncated enough?

Dandandan commented:
I think it’s fine to support `like` for now and leave the further combination / optimization for future work. I see that only simplifying to `starts_with` won't get all the benefits.

Of course, the pruning needs to be correct :). We could add (negative) test cases / add issues if the already implemented logic for `=` pruning is incorrect.

adriangb commented:
Well there's no tests for hypothetical cases with truncated stats. All of the tests are against the stats themselves.

The current implementation is certainly not wrong in the real world, but I'm not sure if that's because it's not used in situations where stats are truncated, if truncation only happens at extremes like a 10MB value where practically it's not a problem, etc.

alamb commented:
I suggest we (I can help tomorrow):
1. File a ticket to simplify `like` to `=` and `starts_with` when possible (to help follow on optimizations like this)
2. File / find a ticket about testing with truncated statistics
3. Determining what, if anything, is left to add directly to pruning prediate

adriangb commented:
Sounds good.

All of that said, I think this PR is currently as correct as `=` and had pretty good test coverage. Does it need to wait on those tasks or can it proceed in parallel?

findepi commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1610,6 +1624,79 @@ fn build_statistics_expr(
+    let (min_lit, max_lit) = if let Some(wildcard_index) = first_wildcard_index {
+        let prefix = &s[..wildcard_index];
+        let prefix_min_lit = Arc::new(phys_expr::Literal::new(ScalarValue::Utf8(Some(
+            format!("{prefix}\u{10ffff}"),
```
When implementing similar thing for Trino (https://github.com/trinodb/trino/commit/6aea881752ef9a83521a4a20bccf510bf9a31d2c), i decided to stay within ASCII characters to avoid any potential issues due to misinterpretation of "difficult" code points.

adriangb replied:
Hmm my intuition was that you want to add the highest character to the upper lower bound of the min value such that `'A%'` can match `'AB'`. Assuming a column with only 1 row `"AB"` and the query `'AB' like 'A%'`:
- `'AB' <= 'A\u{10ffff}' and 'A' <= 'AB'` -> `t`
- `'AB' <= 'A' and 'A\u{10ffff}' <= 'AB'` -> `f`
Right?

adriangb replied:
What does it mean to `stay within ASCII`? As far as I know everything here is Utf8 so I'm not sure how we can restrict it to ASCII?

adriangb replied:
I'm wondering if comment is related to the confusing naming in https://github.com/apache/datafusion/pull/12978/files#r1810513072?

findepi replied:
> What does it mean to `stay within ASCII`? As far as I know everything here is Utf8 so I'm not sure how we can restrict it to ASCII?

We can if only we want to. The code will do whatever we ask it to do.
Then the question is whether we want to. If we apply the predicate locally in memory only, then no need to be cautious, no need for "stay within ASCII". If we later interop with other systems (eg send a plan somewhere or TableProvider calls remote system), then it might be beneficial to restrict ourselves.





> Hmm my intuition was that you want to add the highest character to the upper lower bound

i agree with this

> ... of the min value

min value of what?
all column values need to be in the range `[like_constant_prefix, like_constant_prefix[0..-1] + \u10ffff)`

findepi replied:
> > ... of the min value
> 
> min value of what?

i get it now

So for a column we have stats: min value and max value. Let's call them `col_min` and `col_max`.
For like `AB%` we derive lower and upper bound (`AB` and `AB\u10ffff` which is actually incorrect, will comment about this elsewhere).

For pruning we need to check whether `[col_min, col_max] ∩ [lower_bound, upper_bound)` is non-empty (note the upper_bound will be non-inclusive)
It's empty when `upper_bound <= col_min OR  col_max < lower_bound`
It's non-empty when `upper_bound > col_min AND col_max >= lower_bound`

findepi replied:
For correct upper bound and why exclusive see https://github.com/apache/datafusion/pull/12978#discussion_r1816614140

findepi commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1610,6 +1624,79 @@ fn build_statistics_expr(
+    let scalar_expr = expr_builder.scalar_expr();
+    // check that the scalar is a string literal
+    let s = extract_string_literal(scalar_expr)?;
+    // **IMPORTANT** we need to make sure that the min and max are in the range of the prefix
+    // If we truncate 'A%' to 'A', we need to make sure that 'A' is less than 'AB' so that
+    // when we make this a range query we get 'AB' <= 'A\u{10ffff}' AND 'A' <= 'AB'.
+    // Otherwise 'AB' <= 'A' AND 'A' <= 'AB' would be *wrong* because 'AB' LIKE 'A%' is should be true!
```
This important, but a bit difficult to follow.

adriangb replied:
I'm open to clarifications on the wording. The point I'm trying to make is why we have to append characters.

findepi replied:
I'd remove this whole comment. If someone understand how LIKE works, they will get the code without comment.
If someone doesn't understand how LIKE works, they won't understand the code even with the comment.

findepi commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1610,6 +1624,79 @@ fn build_statistics_expr(
+    let min_expr = Arc::new(phys_expr::BinaryExpr::new(
+        min_column_expr.clone(),
+        Operator::LtEq,
+        min_lit,
```
`min_lit` actually represents max value (upper bound)
would you consider swapping the naming of the variables?

Also, the upper bound has added "Z" ('highest codepoint') at the end, so can be compared with Lt without Eq part

adriangb replied:
> min_lit actually represents max value (upper bound)
> would you consider swapping the naming of the variables?

I'm open to any suggestions on naming but I do think it is confusing because `min_lit` is the _upper bound_ on `col_min` and `max_lit` is the lower bound on `col_max` 🤯 

> Also, the upper bound has added "Z" ('highest codepoint') at the end, so can be compared with Lt without Eq part

👍🏻

findepi replied:
> it is confusing because `min_lit` is the _upper bound_ on `col_min` and `max_lit` is the lower bound on `col_max` 🤯

Yes, it is.

i understand now that you're fitting this into terminology of existing code.
i am not sure what the right naming would be. maybe neutral: lower_bound and upper_bound?

findepi commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1610,6 +1624,79 @@ fn build_statistics_expr(
+        ))));
+        (prefix_min_lit, prefix_max_lit)
+    } else {
+        let prefix_lit =
+            Arc::new(phys_expr::Literal::new(ScalarValue::Utf8(Some(s.clone()))));
+        (prefix_lit.clone(), prefix_lit)
```
In such case we should produce single Eq check

adriangb replied:
Not sure what you mean by a single eq check. We are basically saying `col like 'constant'` -> `col = 'constant'` -> `col_min <= 'constant' and 'constant' <= col_max`

adriangb commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1610,6 +1624,79 @@ fn build_statistics_expr(
+    // Otherwise 'AB' <= 'A' AND 'A' <= 'AB' would be *wrong* because 'AB' LIKE 'A%' is should be true!
+    // Credit to https://stackoverflow.com/a/35881551 for inspiration on this approach.
+    // ANSI SQL specifies two wildcards: % and _. % matches zero or more characters, _ matches exactly one character.
+    let first_wildcard_index = s.find(['%', '_']);
```
```
> SELECT '' LIKE '' ESCAPE '%';
Execution error: LIKE does not support escape_char
```

So I think not.
But just for reference, how would you suggest escape characters be handled? I've never used them in practice (I think at that point I'd just go for a regex).

findepi commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1610,6 +1624,79 @@ fn build_statistics_expr(
+    let s = extract_string_literal(scalar_expr)?;
+    // **IMPORTANT** we need to make sure that the min and max are in the range of the prefix
+    // If we truncate 'A%' to 'A', we need to make sure that 'A' is less than 'AB' so that
+    // when we make this a range query we get 'AB' <= 'A\u{10ffff}' AND 'A' <= 'AB'.
```
for `A%` pattern the lower bound is `A` (obvious)
what should be the upper bound?

`A\u{10ffff}` is not a correct upper bound since `A\u{10ffff}\u{10ffff}` is even bigger but still matches `A%` input.
The correct upper bound would be:
- `A\u{10ffff}\u{10ffff}\u{10ffff}...\u{10ffff}`  include -- up to max length of the column, so potentially very very long, so absolutely not practical
- `B` (exclusive).

Thus to calculate upper bound you need (pseudo-code)

```rust
let s = extract_string_literal(scalar_expr)?;
let first_wildcard_index = ...;
let prefix = &s[..wildcard_index];
let last_incrementable_character = /* find last code point of `prefix` that can be incremented
   if we choose to stay within ascii, this will be a code point < 127
   otherwise it will be any code point != the max code point (0x10FFFF) */;
if last_incrementable_character not found {
  // For `%`, or `\u{10ffff}...\u{10ffff}%` patterns, we cannot calculate an upper bound
  return None
}
let upper_bound = 
   prefix[..last_incrementable_character-1] +  // take prefix of the prefix up to  and excluding the last character that can be incremented
   str(prefix[last_incrementable_character] + 1) // take last character and increment it
```

adriangb replied:
I'm a bit confused about this explanation. Maybe you can provide a failing test case that would help me understand? There is already a test case for `'A%'` and it is as far as I can tell doing the correct thing.

findepi replied:
try add a test case for `'A%'` where stats are `min=AB` `max=A\u{10ffff}\u{10ffff}\u{10ffff}`

adriangb replied:
Thanks for the suggestion. I added a test case in https://github.com/apache/datafusion/pull/12978/commits/e29ed50442505cceb0c3e1de118baee395de1368. It worked as expected, let me know if I got the expected outcomes wrong or missed something.

findepi replied:
See https://github.com/apache/datafusion/pull/12978#discussion_r1819043402 on how to make the test expose the problem

findepi commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1610,6 +1624,79 @@ fn build_statistics_expr(
+    // If we truncate 'A%' to 'A', we need to make sure that 'A' is less than 'AB' so that
+    // when we make this a range query we get 'AB' <= 'A\u{10ffff}' AND 'A' <= 'AB'.
+    // Otherwise 'AB' <= 'A' AND 'A' <= 'AB' would be *wrong* because 'AB' LIKE 'A%' is should be true!
+    // Credit to https://stackoverflow.com/a/35881551 for inspiration on this approach.
```
This link isn't useful. That page conveniently avoids any details that are important. Please remove the link.

findepi replied:
I am just not finding this link useful being being inspirational. But source of inspiration doesn't need to be reflected in the code comment 🙂 
I am not asking for replacing if any other link. 

I **personally** find https://github.com/trinodb/trino/commit/6aea881752ef9a83521a4a20bccf510bf9a31d2c valuable because (1) I know this code as its author and (2) it might actually be correct. Once we get the code here correct, that link wouldn't be useful either.

adriangb commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1610,6 +1624,79 @@ fn build_statistics_expr(
+    let scalar_expr = expr_builder.scalar_expr();
+    // check that the scalar is a string literal
+    let s = extract_string_literal(scalar_expr)?;
+    // **IMPORTANT** we need to make sure that the min and max are in the range of the prefix
+    // If we truncate 'A%' to 'A', we need to make sure that 'A' is less than 'AB' so that
+    // when we make this a range query we get 'AB' <= 'A\u{10ffff}' AND 'A' <= 'AB'.
+    // Otherwise 'AB' <= 'A' AND 'A' <= 'AB' would be *wrong* because 'AB' LIKE 'A%' is should be true!
+    // Credit to https://stackoverflow.com/a/35881551 for inspiration on this approach.
```
```suggestion
```

adriangb commented:
@alamb I re-arranged some of the comments on assertions in https://github.com/apache/datafusion/pull/12978/commits/ae3426da4b19ff9e8adf038547b6a4e552190f1f which I feel like helped a lot with readability of the tests. There's a couple other tests with a similar pattern that I think could benefit.

I was also thinking about doing some more black-box testing: I think given any `min, max` you can always convert that into a  RecordBatch with an array of the form `[min, max]` and you should never have the pruning say the array should be excluded but the array has any matches. Does that make sense? Maybe this could even be fuzz tested?

findepi commented:
> I was also thinking about doing some more black-box testing: I think given any `min, max` you can always convert that into a RecordBatch with an array of the form `[min, max]` and you should never have the pruning say the array should be excluded but the array has any matches. Does that make sense? Maybe this could even be fuzz tested?

We could also randomly generate an array with an element or elements within the min max range.
This would cover not only logic at the edges, but also for the values inside the range.

findepi commented:
Some sqllogictests failed and this is expected. @adriangb can you please update them with
```
cargo test --test sqllogictests -- --complete
```
?

findepi commented:
> Some sqllogictests failed and this is expected. @adriangb can you please update them with
> 
> ```
> cargo test --test sqllogictests -- --complete
> ```
> 

You will need to merge with `main` to reproduce the failures (the failing tests are new).
When you do so, you should get a diff like this

```diff
@@ -510,7 +510,7 @@ physical_plan
 01)CoalesceBatchesExec: target_batch_size=8192
 02)--FilterExec: binary_col@0 LIKE %a% AND largebinary_col@1 LIKE %a% AND binaryview_col@2 LIKE %a%
 03)----RepartitionExec: partitioning=RoundRobinBatch(2), input_partitions=1
-04)------ParquetExec: file_groups={1 group: [[WORKSPACE_ROOT/datafusion/sqllogictest/test_files/scratch/parquet/binary_as_string.parquet]]}, projection=[binary_col, largebinary_col, binaryview_col], predicate=binary_col@0 LIKE %a% AND largebinary_col@1 LIKE %a% AND binaryview_col@2 LIKE %a%
+04)------ParquetExec: file_groups={1 group: [[WORKSPACE_ROOT/datafusion/sqllogictest/test_files/scratch/parquet/binary_as_string.parquet]]}, projection=[binary_col, largebinary_col, binaryview_col], predicate=binary_col@0 LIKE %a% AND largebinary_col@1 LIKE %a% AND binaryview_col@2 LIKE %a%, pruning_predicate=CASE WHEN binary_col_null_count@2 = binary_col_row_count@3 THEN false ELSE binary_col_min@0 <= 􏿿 AND  <= binary_col_max@1 END AND CASE WHEN largebinary_col_null_count@6 = largebinary_col_row_count@7 THEN false ELSE largebinary_col_min@4 <= 􏿿 AND  <= largebinary_col_max@5 END AND CASE WHEN binaryview_col_null_count@10 = binaryview_col_row_count@11 THEN false ELSE binaryview_col_min@8 <= 􏿿 AND  <= binaryview_col_max@9 END, required_guarantees=[]
```

thoughts

- for a predicate `LIKE %a%` we should not generate additional conditions at all. There is no min,max predicate we can practically derive from `LIKE %a%`  expression.
- `􏿿`  character is problematic to read and to work with. Some tools don't handle it properly. For example when pasting into iterm with python3 running for inspection, it got displaed as `\U+0FFFF` even though this is code point`x10FFFF`.

findepi commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1610,6 +1624,74 @@ fn build_statistics_expr(
+            ScalarValue::LargeUtf8(Some(s)) => Ok(s),
+            ScalarValue::Utf8View(Some(s)) => Ok(s),
+            ScalarValue::Dictionary(_, value) => unpack_string(value),
+            _ => plan_err!("LIKE expression must be a string literal"),
```
```suggestion
            _ => plan_err!("LIKE pattern literal must be a string"),
```

findepi commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1610,6 +1624,74 @@ fn build_statistics_expr(
+fn build_like_match(
+    expr_builder: &mut PruningExpressionBuilder,
+) -> Result<Arc<dyn PhysicalExpr>> {
+    // column LIKE literal => (min, max) LIKE literal split at % => min <= split literal && split literal <= max
+    // column LIKE 'foo%' => min <= 'foo' && 'foo' <= max
+    // column LIKE '%foo' => min <= '' && '' <= max => true
+    // column LIKE '%foo%' => min <= '' && '' <= max => true
+    // column LIKE 'foo' => min <= 'foo' && 'foo' <= max
```
Turn this into function documentation and place above the function declaration

```
/// Convert `column LIKE literal` where P is a constant prefix of the literal
/// to a range check on the column: `P <= column && column < P'`, where P' is the
/// lowest string after all P* strings.
```

findepi commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -3443,6 +3525,385 @@ mod tests {
+                    Some(""),
+                    Some(""),
+                    Some("AB"),
+                    Some("A\u{10ffff}"),
```
```suggestion
                    Some("A\u{10ffff}\u{10ffff}"),
```

findepi replied:
This will expose the problem we're talking about here https://github.com/apache/datafusion/pull/12978#discussion_r1817937062

adriangb replied:
Thank you, this is great! I believe I fixed it in https://github.com/apache/datafusion/pull/12978/commits/cc0b70b8af42502a4feea0d7b544c7b1f81fafda 😄

adriangb commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1610,6 +1624,74 @@ fn build_statistics_expr(
+    }
+
+    fn extract_string_literal(expr: &Arc<dyn PhysicalExpr>) -> Result<&String> {
+        if let Some(lit) = expr.as_any().downcast_ref::<phys_expr::Literal>() {
+            let s = unpack_string(lit.value())?;
+            return Ok(s);
+        }
+        plan_err!("LIKE expression must be a string literal")
```
And we couldn't handle those regardless, there's no way to push them down to statistics. So I think only handling Literal is okay. Liter

findepi replied:
Yes, we cannot handle them, but we should improve error message. As in suggestions.

adriangb replied:
These errors don't get surfaced to users. They get handled within this module, becoming `true` by default (and multiple `true` get coalesced). It's the same thing that would happen if you put in a clause like `AND my_special_function(column)`.

findepi replied:
I know they aren't. The exception-driven control flow is not generally recommended.
Let's reserve err result for actual errors, if possible.

adriangb replied:
Okay I can change them to options. I was just following the pattern followed elsewhere, plus the errors are a convenient place to put a string explaining why we are bailing as an fyi to other devs.

adriangb replied:
Changed to options in https://github.com/apache/datafusion/pull/12978/commits/5bcca3ef98def407822d23318ac8b6584f9f8c24

adriangb commented:
> > Some sqllogictests failed and this is expected. @adriangb can you please update them with
> > ```
> > cargo test --test sqllogictests -- --complete
> > ```
> 
> You will need to merge with `main` to reproduce the failures (the failing tests are new). When you do so, you should get a diff like this

Thanks I was wondering how to update that. When I run it I get various errors along the lines of:
```
Execution error: Error completing "string/string_view.slt": failed to rename file from test_files/string/init_data.slt.part.temp to test_files/string/./init_data.slt.part
Execution error: Error completing "string/large_string.slt": failed to rename file from test_files/string/string_query.slt.part.temp to test_files/string/./string_query.slt.part
Execution error: Error completing "string/dictionary_utf8.slt": failed to rename file from test_files/string/string_query.slt.part.temp to test_files/string/./string_query.slt.part
```

Any idea what's going on?

adriangb commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -3577,7 +3585,11 @@ mod tests {
         // s1 ["", "A"]  ==> some rows could pass (must keep)
         // s1 ["", ""]  ==> no rows can pass (not keep)
-        let expected_ret = &[true, true, false, false, true, true, true, false];
+        // s1 ["AB", "A\u{10ffff}"]  ==> no rows can pass (not keep)
```
I take it you are saying each one of these should be their own test right? I agree that would make it a lot clear. It'd be more boilerplate code as well but IMO it'd be worth it. I was just following the pattern other tests use, and this started as a couple test cases and has grown considerably...

findepi replied:
> I take it you are saying each one of these should be their own test right?

not necessarily separate test funciton

just that having test case input and expected output next to each other would help readability.
currently the test workarounds this by replicating input inside code comments.

> just following the pattern other tests use

👍 
let's not change this within this PR

findepi commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1648,22 +1649,46 @@ fn build_like_match(
+        let mut data = data.as_bytes().to_vec();
+        for idx in (0..data.len()).rev() {
+            let original = data[idx];
+            let (byte, overflow) = original.overflowing_add(1);
```
incrementing character 127 should be supported.
same with other utf-8 byte sequences where last byte is saturated.

let's operate on code points, not bytes.

adriangb replied:
Changed to operate on code points in https://github.com/apache/datafusion/pull/12978/commits/5bcca3ef98def407822d23318ac8b6584f9f8c24

findepi commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1648,22 +1649,46 @@ fn build_like_match(
+    /// Try and increment the the string's bytes from right to left, returning when the result
+    /// is a valid UTF8 string. Returns `None` when it can't increment any byte.
+    /// Vendored from https://github.com/apache/arrow-rs/blob/56525efbd5f37b89d1b56aa51709cab9f81bc89e/parquet/src/column/writer/mod.rs#L1432-L1448
+    fn increment_utf8(data: &str) -> Result<String> {
```
define the function outside of `build_like_match` and add unit tests for it.

adriangb replied:
done

findepi commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1610,6 +1625,100 @@ fn build_statistics_expr(
+    }
+    let (min_lit, max_lit) = if let Some(wildcard_index) = first_wildcard_index {
+        let prefix = &s[..wildcard_index];
+        let prefix_min_lit = Arc::new(phys_expr::Literal::new(ScalarValue::Utf8(Some(
```
please rename these min_lit, max_lit to upper_bound and lower_bound 
because their meaning is reversed.

```
let (upper_bound, lower_bound) = 
```

Also, it would be good to extract like `literal pattern -> (lower_bound, upper_bound)` function and add unit tests for it.

adriangb replied:
I did the rename for now let's see if that's good enough

adriangb commented:
> Yes -- #12752
> 
> but don't worry. see [#12978 (comment)](https://github.com/apache/datafusion/pull/12978#issuecomment-2441489968) . this test case shouldn't change after all

Having other issues now. I tried in 46531f5bf7be7a59255fc8723f0be3efd3beb908 but got a bunch of errors from missing files. I tried to follow what CI is doing but got to compiling the databricks repo and got a missing `malloc.h` so I stopped there since I wasn't even sure if I was barking up the right tree. Is this the expected workflow for updating these snapshots? It might be helpful if failed CI runs offered them as a downloadable artifact.

adriangb commented:
> for a predicate LIKE %a% we should not generate additional conditions at all. There is no min,max predicate we can practically derive from LIKE %a% expression

By the way I added a special case for the wildcard being the first character (no prefix) to bail out and add no filtering

adriangb commented:
Looks like the special case on `'%...'` actually eliminated the diff in those tests, so no need to help me figure out how to update them now. I'll re-request review from both of you. Thank you for your help so far.

alamb commented:
This is on my list to review today

alamb commented:
I didn't make it here today -- it is the top of my list for tomorrow

alamb reviewed (dismissed):
Thank you @adriangb and @findepi

In my opinion this PR is quite close -- the only thing I am not sure about is test coverage for `increment_utf8`

I think we could potentially do that as a follow on PR. I also left a bunch of comments that might be able to improve the code, but I don't think they are necessary

But really all in all this is very clever 👏 

> > I was also thinking about doing some more black-box testing: I think given any `min, max` you can always convert that into a RecordBatch with an array of the form `[min, max]` and you should never have the pruning say the array should be excluded but the array has any matches. Does that make sense? Maybe this could even be fuzz tested?
> 
> We could also randomly generate an array with an element or elements within the min max range. This would cover not only logic at the edges, but also for the values inside the range.

Yes, I agree that a black box fuzz tester would be most useful for this area.

adriangb commented:
Thanks for the review Andrew. Let's open an issue to track a fuzz tester for predicate pushdown.

findepi commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1610,6 +1629,126 @@ fn build_statistics_expr(
+fn build_like_match(
+    expr_builder: &mut PruningExpressionBuilder,
+) -> Option<Arc<dyn PhysicalExpr>> {
+    // column LIKE literal => (min, max) LIKE literal split at % => min <= split literal && split literal <= max
+    // column LIKE 'foo%' => min <= 'foo' && 'foo' <= max
+    // column LIKE '%foo' => min <= '' && '' <= max => true
+    // column LIKE '%foo%' => min <= '' && '' <= max => true
+    // column LIKE 'foo' => min <= 'foo' && 'foo' <= max
```
in https://github.com/apache/datafusion/pull/12978#discussion_r1819023885 it was proposed to reword this and make a function doc comment.

If you want to keep as is, please make sure `=>` for implication and those comparisons can be easily read - maybe add braces?

findepi commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1567,6 +1581,11 @@ fn build_statistics_expr(
             ))
         }
+        Operator::LikeMatch => build_like_match(expr_builder).ok_or_else(|| {
+            plan_datafusion_err!(
+                "LIKE expression with wildcard at the beginning is not supported"
+            )
```
The idea to return option from `build_like_match` was to avoid communication "i can't derive predicate" via errors.
Changing `build_statistics_expr` to return Option is too much for this PR, but we can maybe return an always-true expression?

adriangb replied:
That's what `unhandled_hook` does. It won't always be a literal `true`. I think it's easiest to follow the pattern in the rest of this file for now and let it bubble up and hit `unhandled_hook`.

findepi commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1610,6 +1629,126 @@ fn build_statistics_expr(
+    }
+    let (upper_bound, lower_bound) = if let Some(wildcard_index) = first_wildcard_index {
+        let prefix = &s[..wildcard_index];
+        let upper_bound_lit = Arc::new(phys_expr::Literal::new(ScalarValue::Utf8(Some(
+            increment_utf8(prefix)?,
+        ))));
+        let lower_bound_lit = Arc::new(phys_expr::Literal::new(ScalarValue::Utf8(Some(
+            prefix.to_string(),
+        ))));
+        (upper_bound_lit, lower_bound_lit)
```
When dealing with bounds or ranges, it's  customary to give them in `(low, high)` order:

```suggestion
        let lower_bound_lit = Arc::new(phys_expr::Literal::new(ScalarValue::Utf8(Some(
            prefix.to_string(),
        ))));
        let upper_bound_lit = Arc::new(phys_expr::Literal::new(ScalarValue::Utf8(Some(
            increment_utf8(prefix)?,
        ))));
        (lower_bound_lit, upper_bound_lit)
```

findepi replied:
nit: can you perhaps compute and return lower bound first?

adriangb replied:
done thank you

findepi commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1610,6 +1629,126 @@ fn build_statistics_expr(
+        // there's no filtering we could possibly do, return an error and have this be handled by the unhandled hook
+        return None;
+    }
+    let (upper_bound, lower_bound) = if let Some(wildcard_index) = first_wildcard_index {
```
nice!

adriangb replied:
We do want to handle the case where there is no wildcard but the string is not empty. E.g. `col like 'a'`. So I don't think this will be as simple: if we get None we actually have to check for `col like 'a'`.

findepi commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1610,6 +1629,126 @@ fn build_statistics_expr(
+/// Increment a UTF8 string by one, returning `None` if it can't be incremented.
+/// This makes it so that the returned string will always compare greater than the input string
+/// or any other string with the same prefix.
+/// E.g. `increment_utf8("foo") >= "foo"` and `increment_utf8("foo") >= "fooz"`
```
Nice example.
This could also say that `increment_utf8("foo")` returns "fop"

adriangb replied:
```suggestion
/// E.g. `increment_utf8("foo") >= "foo"` and `increment_utf8("foo") >= "fooz"`
/// In this example `infrement_utf8("foo") == "fop"
```

findepi commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1610,6 +1629,126 @@ fn build_statistics_expr(
+    fn is_valid_unicode(c: char) -> bool {
+        let cp = c as u32;
+
+        // Filter out Unicode surrogate pairs range
+        if (0xD800..=0xDFFF).contains(&cp) {
+            return false;
+        }
+
+        // Filter out non-characters
+        if (0xFDD0..=0xFDEF).contains(&cp) {
+            return false;
+        }
+
+        // Filter out private use areas and other invalid ranges
+        match cp {
+            0xFFFE | 0xFFFF => false,
+            _ if cp >= 0x110000 => false,
+            _ => true,
+        }
```
There should be helpers to inspect a char in Rust sdk or icu.

Or we can "stay within ASCII" to keep things simple for now.

findepi replied:
After applying the above suggestions, this can be

```suggestion
        // Filter out private use area
        if cp >= 0x110000 {
            return false;
        }

        true
```

findepi commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1610,6 +1629,126 @@ fn build_statistics_expr(
+    fn is_valid_unicode(c: char) -> bool {
+        let cp = c as u32;
+
+        // Filter out Unicode surrogate pairs range
+        if (0xD800..=0xDFFF).contains(&cp) {
+            return false;
+        }
```
According to https://doc.rust-lang.org/std/primitive.char.html, this is redundant. `char` cannot be a code point within surrogate range.

```suggestion
```

findepi commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1610,6 +1629,126 @@ fn build_statistics_expr(
+            return false;
+        }
+
+        // Filter out non-characters
+        if (0xFDD0..=0xFDEF).contains(&cp) {
+            return false;
+        }
```
Per https://www.unicode.org/versions/corrigendum9.html, 0xFFFE | 0xFFFF are also non-characters

```suggestion
        // Filter out non-characters (https://www.unicode.org/versions/corrigendum9.html)
        if [0xFFFE, 0xFFFF].contains(&cp) || (0xFDD0..=0xFDEF).contains(&cp) {
            return false;
        }
```

findepi reviewed (approved):
Thank you!

adriangb commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1610,6 +1629,126 @@ fn build_statistics_expr(
+
+/// Increment a UTF8 string by one, returning `None` if it can't be incremented.
+/// This makes it so that the returned string will always compare greater than the input string
+/// or any other string with the same prefix.
```
```suggestion
/// or any other string with the same prefix.
/// This is necessary since the statistics may have been truncated: if we have a min statistic
/// of "fo" that may have originally been "foz" or anything else with the prefix "fo".
```

adriangb commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1610,6 +1629,126 @@ fn build_statistics_expr(
+    let (upper_bound, lower_bound) = if let Some(wildcard_index) = first_wildcard_index {
+        let prefix = &s[..wildcard_index];
+        let upper_bound_lit = Arc::new(phys_expr::Literal::new(ScalarValue::Utf8(Some(
+            increment_utf8(prefix)?,
```
I added a comment.

adriangb commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1610,6 +1629,126 @@ fn build_statistics_expr(
+/// This makes it so that the returned string will always compare greater than the input string
+/// or any other string with the same prefix.
+/// E.g. `increment_utf8("foo") >= "foo"` and `increment_utf8("foo") >= "fooz"`
+fn increment_utf8(data: &str) -> Option<String> {
```
Not really, it's something that @davidhewitt and I cooked up. I don't think we consider ourselves experts in the utf8 spec either. Luckily @findepi 's comments simplified it a bit 😃.

adriangb replied:
I know there's an implementation in the string truncation for stats. This code was originally based on that  (copied in) but we modified it as part of input from @findepi.

etseidl replied:
Also not a UTF-8 expert 😛.  The code @alamb is referring to is [here](https://github.com/rapidsai/cudf/blob/1b82963df736f3ad71b003443a4de1414f3ce2e5/cpp/src/io/parquet/page_enc.cu#L2759), and unit test is [here](https://github.com/rapidsai/cudf/blob/1b82963df736f3ad71b003443a4de1414f3ce2e5/cpp/tests/io/parquet_writer_test.cpp#L882). I had to do it this way because STL isn't usable in code that runs on the GPU, and I couldn't handle cases where incrementing the character increases the byte count.

Here we have the luxury of being able to convert the encoded bytes into a u32 representation of the code point, increment that and then re-encode. Much nicer :) The testing of the increment is pretty thorough...U+7F to U+80 is the tricky one. U+00FF to U+0100 would be another to test as that also overflows the lower 6 bits of a 2-byte character.

Looks pretty sound to me. 👍

adriangb commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1610,6 +1629,126 @@ fn build_statistics_expr(
+
+        // Try incrementing the code point
+        if let Some(next_char) = char::from_u32(original + 1) {
+            // Check if it's a valid continuation
```
```suggestion
```

adriangb commented:
@alamb @findepi I think I've accepted or applied all of your excellent feedback, and I added some coverage / more test cases to the `increment_utf8` function 😄

findepi commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1610,6 +1629,124 @@ fn build_statistics_expr(
+fn build_like_match(
+    expr_builder: &mut PruningExpressionBuilder,
+) -> Option<Arc<dyn PhysicalExpr>> {
+    // column LIKE literal => (min, max) LIKE literal split at % => min <= split literal && split literal <= max
+    // column LIKE 'foo%' => min <= 'foo' && 'foo' <= max
+    // column LIKE '%foo' => min <= '' && '' <= max => true
+    // column LIKE '%foo%' => min <= '' && '' <= max => true
+    // column LIKE 'foo' => min <= 'foo' && 'foo' <= max
```
see https://github.com/apache/datafusion/pull/12978#discussion_r1819023885  https://github.com/apache/datafusion/pull/12978#discussion_r1823893452

there could be a function doc comment

```
/// Convert `column LIKE literal` where P is a constant prefix of the literal
/// to a range check on the column: `P <= column && column < P'`, where P' is the
/// lowest string after all P* strings.
```

adriangb replied:
added

alamb commented:
This is one of the PRs I hope to push through this weekend -- what I would really like to do given the potential for very subtle bugs in this code, is to write some more tests. I keep hoping I will find time, but I haven't. Maybe this weekend will be better

findepi commented:
Given LIKE can get desugared to something else during planning (eg  https://github.com/apache/datafusion/pull/13061), maybe it would be great to support this functionality with end-to-end tests.

adriangb commented:
> Given LIKE can get desugared to something else during planning (eg #13061), maybe it would be great to support this functionality with end-to-end tests.

I feel like we need both e2e and more focused tests: e2e tests risk depending on a combination of the optimizer and pruning rewrites that could be broken if e.g. someone customized the optimizer (which I believe is pluggable?).

drauschenbach commented:
This PR should possibly indicate that it Closes #507.

alamb commented:
> This PR should possibly indicate that it Closes #507.

Updated the description

alamb reviewed (commented):
Sorry for the delay in merging this one, just the pruning logic is so subtle I don't want to introduce any unintended behaviors. 

I wrote up some fuzz tests here: https://github.com/apache/datafusion/pull/13253

They don't yet pass -- I think there might be some bug in the test. I need to spend more time debugging it but I am out of time for today

adriangb commented:
No worries, I'll try to take a look at #13253. I agree it's better to be cautious and make sure we get this right rather than create a bug that results in incorrect data.

findepi commented:
> I wrote up some fuzz tests here: #13253
> 
> They don't yet pass -- I think there might be some bug in the test.

Good tests should find https://github.com/apache/datafusion/issues/12637 and therefore shouldn't pass.

alamb commented:
To follow up here -- what I think this PR needs prior to merging is to add fuzz testing as I think it has the potential to have hard to track down / subtle bugs

Some important corner cases:
1. With truncated statistics (actual min/max different than stats)
2. With truncated prefix matching -- as in `min` / `max` is actually `foo` but the query has a predicate like `column='fo'`

etseidl commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -1610,6 +1629,127 @@ fn build_statistics_expr(
+    fn is_valid_unicode(c: char) -> bool {
+        let cp = c as u32;
+
+        // Filter out non-characters (https://www.unicode.org/versions/corrigendum9.html)
+        if [0xFFFE, 0xFFFF].contains(&cp) || (0xFDD0..=0xFDEF).contains(&cp) {
+            return false;
+        }
```
After reading the link, I'm not sure we need to filter out non-characters. They are allowed to appear in well formed strings, and will sort properly AFAICT. Then this whole function could boil down to
```rust
  c as u32 < 0x110000
```

findepi replied:
From the link above

> Noncharacters in the Unicode Standard are intended for internal use and have no standard interpretation when exchanged outside the context of internal use.

whereas the filter condition we gonna produce here may be exchanged between systems

findepi replied:
Usually is, but AFAICT it can also reach `TableProvider.scan` and then ... 🤷‍♂️

findepi commented on datafusion/core/src/physical_optimizer/pruning.rs:
```diff
@@ -3443,6 +3583,425 @@ mod tests {
+        // Test that if we can't increment the last character we do the previous one and truncate
+        assert_eq!(increment_utf8("a\u{10FFFF}").unwrap(), "b");
+
+        // Test surrogate pair range (0xD800..=0xDFFF)
+        assert_eq!(increment_utf8("a\u{D7FF}").unwrap(), "b");
+        assert!(increment_utf8("\u{D7FF}").is_none());
```
why so? 0xD7FF + 1 would be 0xD800, not 0xE800

alamb replied:
BTW @etseidl has a PR upstream in arrow-rs related to incrementing utf8:
-  https://github.com/apache/arrow-rs/pull/6870

Maybe we can focus our efforts on making sure we have a solid utf8 increment routine and then use it in both places

findepi replied:
thanks for the link.



@alamb i am not against reusing, but note that business logic for Parquet stats and here is different.
For stats, it's enough to find _some_ value bigger than input.
Here, it's mandatory to find _next_ value bigger than input.


i have feeling that the routine in this PR is ready to be merged.
@adriangb can you please rebase the PR?
if a better version emerges, it's not a problem to improve later, but there is no need to tie this PR

adriangb replied:
@findepi  @alamb [feels that this needs some fuzz testing before it can be merged](https://github.com/apache/datafusion/pull/12978#issuecomment-2504619520) and I heavily agree.

I plan on doing this but have not had time.
Since we have a DataFusion meetup next week in Chicago I've asked for 2 days of dedicated work to spend on this + prepping for the meetup next week. So hopefully I can tackle that Monday and Tuesday.

etseidl replied:
> why so? 0xD7FF + 1 would be 0xD800, not 0xE800

Because 0xD800-0xDFFF are not valid code points. The next higher valid code point is 0xE000. It's a happy coincidence that this happens.

adriangb commented:
Hi @alamb I took a stab at fuzz tests in 969e83c82.
They're heavy and slow so I had to restrict the search space a lot more than I would have liked. Maybe you or @tustvold can suggest ways to cut out the heavy parsing of Parquet metadata and such to speed these up? Ultimately I do think it's worth re-using whatever creates parquet stats from data so that we use the "real" thing but I don't think we need to test the serialization / deserialization repeatedly like this does.
Also happy to restrict the search space by being more deliberate about how we build the values, row groups, predicates, etc.

alamb commented:
Thanks @adriangb  -- I have put this PR on my review queue and will attempt to re-review it carefully tomorrow

alamb commented:
Clearly I failed to review this -- I will do so hopefully later today but may be tomorrow

alamb commented:
This is still on my list, hopefully other people can check it out too

alamb commented:
This is my top priority after DF 44 is released:
- #13334

alamb commented on datafusion/core/tests/fuzz_cases/pruning.rs:
```diff
@@ -0,0 +1,247 @@
+        "~",
+        "ß",
+        "℣",
+        "%", // this one is useful for like/not like tests since it will result in randomly inserted wildcards
```
👍

alamb commented on datafusion/core/tests/fuzz_cases/pruning.rs:
```diff
@@ -0,0 +1,247 @@
+    values.truncate(100);
+
+    let mut row_groups = vec![];
+    // generate all combinations of values for row groups (1 or 2 values per rg, more is unessecarry since we only get min/max stats out)
```
```suggestion
    // generate all combinations of values for row groups (1 or 2 values per rg, more is unnecessary since we only get min/max stats out)
```

alamb commented on datafusion/physical-optimizer/src/pruning.rs:
```diff
@@ -1607,6 +1629,127 @@ fn build_statistics_expr(
+/// of "fo" that may have originally been "foz" or anything else with the prefix "fo".
+/// E.g. `increment_utf8("foo") >= "foo"` and `increment_utf8("foo") >= "fooz"`
+/// In this example `increment_utf8("foo") == "fop"
+fn increment_utf8(data: &str) -> Option<String> {
```
would it be ok to potentially replace this with the implementation from @etseidl  in https://github.com/apache/arrow-rs/pull/6870 ? 

If so, I can file a ticket to do so as a follow on

adriangb replied:
I haven't reviewed that implementation but yes I think we should consider it!

adriangb commented:
> Thanks @adriangb -- I think this PR is ready to go
> 
> One thing I noticed is that the fuzz test takes over a minute on my machine:
> 
> ```
>         SLOW [> 60.000s] datafusion::fuzz fuzz_cases::pruning::test_fuzz_utf8
>         PASS [  65.772s] datafusion::fuzz fuzz_cases::pruning::test_fuzz_utf8
> ------------
>      Summary [  72.749s] 47 tests run: 47 passed (1 slow), 0 skipped
> andrewlamb@Mac:~/Software/datafusion$
> ```
> 
> Is there some way to make it faster? Maybe with multiple threads or crank down the number of things to teset?

Yeah this is what I was hinting at in https://github.com/apache/datafusion/pull/12978#issuecomment-2542335627.

I'm happy to throw threads at it for a start, and restricting the search space might be necessary but I think requires a more careful eye to minimize how much valuable testing is discarded. The other thing that I think we can do is speed up the tests themselves, in particular minimizing unnecessary round trips to Parquet, but I'm not sure where the right places to hook in would be that still give us a realistic test but remove the need to re-parse the same data over and over again.

alamb commented:
> Yeah this is what I was hinting at in https://github.com/apache/datafusion/pull/12978#issuecomment-2542335627.
> 
> I'm happy to throw threads at it for a start, and restricting the search space might be necessary but I think requires a more careful eye to minimize how much valuable testing is discarded. The other thing that I think we can do is speed up the tests themselves, in particular minimizing unnecessary round trips to Parquet, but I'm not sure where the right places to hook in would be that still give us a realistic test but remove the need to re-parse the same data over and over again.


Awesome -- I'll try and find time later today or tomorrow to give it a critical eye. Otherwise I'll plan to merge this PR later today or tomorrow as well.

alamb commented:
> Yeah this is what I was hinting at in [#12978 (comment)](https://github.com/apache/datafusion/pull/12978#issuecomment-2542335627).
> 
> I'm happy to throw threads at it for a start, and restricting the search space might be necessary but I think requires a more careful eye to minimize how much valuable testing is discarded. The other thing that I think we can do is speed up the tests themselves, in particular minimizing unnecessary round trips to Parquet, but I'm not sure where the right places to hook in would be that still give us a realistic test but remove the need to re-parse the same data over and over again.


- Filed https://github.com/apache/datafusion/issues/13946 to track this

alamb commented:
Thank you again @adriangb  for bearing with us -- I know this took a long time

However, I am pretty stoked that we now have this optimization and it is an example of the very careful engineering required for this kind of optimization. The fact we are at this point in DataFusion is pretty sweet in my mind

🚀

alamb commented on datafusion/physical-optimizer/src/pruning.rs:
```diff
@@ -1605,7 +1627,129 @@ fn build_statistics_expr(
+    // column LIKE '%foo%' => min <= '' && '' <= max => true
+    // column LIKE 'foo' => min <= 'foo' && 'foo' <= max
+
+    fn unpack_string(s: &ScalarValue) -> Option<&String> {
```
I proposed pulling this function out into its own API here: https://github.com/apache/datafusion/pull/14167
